enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,209 @@
+
+#### Deleting CRs
+
+The operand cleanup process, simply put, is the removal of all CRs being reconciled by an operator so it can run finalizers for each CR and clean up any associated operator-created resources.
+
+From OLM's perspective these would be all CRs in the `OperatorGroup`'s [target namespaces][target-namespaces] that are of the same type as the [owned APIservices][owned-apiservices] listed in the CSV. This represents the set of all CRs that the operator reconciles and is responsible for cleaning up after.
+
+OLM should not delete CRs of the type [required APIService][required-apiservices] as those are likely created by the operator, and whose cleanup should be handled by the operator's finalizer.
+
+#### Finalizers on CSVs
+
+OLM can initiate the process of operand clean up when the operator is uninstalled, which is when the CSV is deleted. However, deleting a CSV causes the operator deployment (owned by the CSV) to be garbage collected.
+
+So in order to keep the operator around to process CR deletion events, OLM can add a finalizer to the CSV to prevent the CSV and the operator from being removed. Once OLM has completed operand cleanup and can verify the removal of all CRs, the CSV finalizer can be cleared so that the operator can be safely removed.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+metadata:
+  ...
+  finalizers:
+  - olm/operand-cleanup
+```
+
+In the event of operand cleanup being stalled due to some operator error, OLM will not clear the CSV finalizer so that the operator stays alive indefinitely. This way the operator can be observed to figure out the required action to unblock cleanup.
+
+#### Operand cleanup API
+
+The CSV spec can have an optional field `spec.cleanup` which lets a user configure the cleanup behaviour before the CSV is deleted.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+spec:
+  ...
+  cleanup:
+    enabled: false
+```
+
+With `spec.cleanup.enabled: true` OLM will first add the `olm/operand-cleanup` finalizer on the CSV. Now when the CSV is deleted by the user, OLM will delete all CRs for the operator's owned APIs, in the target namespaces of that operator. Once the CRs are successfully removed (after the operator clears its finalizers), OLM will clear the finalizer from the CSV to let the CSV and operator be uninstalled.
+
+With `spec.cleanup.enabled: false` OLM will remove the `olm/operand-cleanup` finalizer if it is already present on the CSV. This allows users to opt-out of cleanup and not have OLM delete any CRs.
```

Comment:
Can you define behavior when:

- `spec.cleanup.enabled` was true
- operator was uninstalled but fails to remove finalizers on a few CRs
- `spec.cleanup.enabled` is set to false
- operator is removed

Would the existing CRs still have finalizers on them?

Reply:
Ah sorry missed this in my last update.

Yes but in my mind the sequence of events would be a little different:
- `spec.cleanup.enabled` was true
- operator was uninstalled but fails to remove finalizers on a few CRs
- `spec.cleanup.enabled` is set to false
- It has no effect and the CSV is still pending deletion
- the operator is still running
- The CRs are still pending deletion with their own finalizers

Cleanup should be toggled before CSV deletion/operator uninstall. Once you initiate uninstall and OLM issues deletes to the CRs, there is no going back.
OLM should never touch the finalizers on the application CRs since that's the operator's job. And removing the operator once cleanup is stalled won't help cleanup.

And if for some reason the user really wants to remove the CSV/operator anyway then that can be done manually by removing the CSV's finalizer. Or deleting the operator.  `spec.cleanup.enabled` is not intended for that.